tools/chpldoc/findUndocumentedSymbols.py

```diff
@@ -0,0 +1,387 @@
+#!/usr/bin/env python3
+
+#
+# Copyright 2020-2023 Hewlett Packard Enterprise Development LP
+# Copyright 2004-2019 Cray Inc.
+# Other additional copyright holders may be indicated within.
+#
+# The entirety of this work is licensed under the Apache License,
+# Version 2.0 (the "License"); you may not use this file except
+# in compliance with the License.
+#
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+This script is used to parse and identify all undocumented symbols in Chapel
+source code. It can be passed any number of Chapel source files or directories,
+which it will recurse looking for Chapel source files by default.
+
+Disclaimer: this script attempts to be as faithful as possible to the
+`chpldoc.cpp` implementation, but may differ slightly in behavior.
+
+The script works on the following symbols
+- Modules
+- Functions
+  - In this sense, a function is anything declared with `proc`, `iter`, or
+    `operator`
+  - This includes functions defined inside of Modules, AggregateDecls, and
+    Interfaces
+  - This will not match nested Functions
+- AggregateDecls
+  - AggregateDecls are Records, Classes, or Unions
+  - This will match nested definitions in Modules and AggregateDecls,
+    e.g. a Record inside a Record
+- Decls
+  - Decls are top level variables defined in Modules or AggregateDecls
+- MultiDecls
+  - Same as Decls, but multiple symbols share the same keyword, e.g.
+    `var a, b = 10;`
+- Enums
+  - Will also match against EnumElements
+  - This will match nested definitions in Modules and AggregateDecls
+- Interfaces
+  - This will match nested definitions in Modules and AggregateDecls
+
+The script pulls out all of these symbols from a parsed source file, and then
+determines if the symbol is documented. A symbol is considered documented if
+the previous sibling node in the AST is a docstring Comment.
+
+Symbols can be marked 'nodoc', meaning it is ok for the docstring to be
+missing. This script considers a symbol to be marked 'nodoc' based on the
+following criteria:
+- If it is explicitly marked with an attribute `@chpldoc.nodoc`
+- If its names is prefixed with `chpl_`
+  - This also checks the `pragma "chpldoc ignore chpl prefix"`
+- If the symbol is marked private
+- If the symbol has a parent that is marked 'nodoc'
```

Comment:
Might help to make it more clear that this is a disjunction, and that not all criteria need to be met at once.